Profiling updates

owaddell opened the pull request:
**Fixes**

* Fix the destination for annotated files when generating a coverage report where some source files are included via relative path.

**Proposed changes**

* Improve coverage report's rendering of annotated source code:
   - Add line numbers.
   - Do not color leading whitespace.
   - Change DOCTYPE to HTML5.
* Add a coverage legend within annotated source files.
* Tidy the HTML output in profile report (and in `swish-test`'s HTML mat output)
* Adapt to a change from version 9.5.1 of Chez Scheme that maintains profile counts even when the collector reclaims source code.
* Strip the common prefix from source files in the coverage summary and sort based on real path.

payneca committed:
profile: update coverage report source code rendering

- add line numbers
- do not color leading whitespace
- change DOCTYPE to HTML5

owaddell-beckman committed:
profile: adapt to csv9.5.1 profiling change

Starting with version 9.5.1, Chez Scheme maintains profile counts even
for code that the collector has reclaimed. We must now explicitly release
those counts by calling profile-release-counters.

Reduce collection overhead by removing the explicit call to collect that
we added in 751076d and let developers decide whether and how to collect
before they call the extended profile:save.

payneca committed:
profile: add coverage legend to source code rendering

payneca committed:
profile: tidy HTML outputs

Ran swish-test's mat report and coverage report through
 - html-tidy 5.8.0 and
 - https://validator.github.io/validator/ version 23.11.29

payneca committed:
profile: add number of sites in each coverage bucket to legend

owaddell-beckman committed:
profile: fix destination for annotated files

If a source file does (include "../../aux/file.ss"), then swish-test
could end up trying to write an annotated file.ss.html outside the
target directory containing the specified coverage summary report.

To fix this we determine a common prefix based on the real paths of the
profiled source files and use the distinct suffix as the path within the
target directory.

The appearance of the coverage report changes slightly since we:
 - sort entries based on the real path of the file,
 - use natural-string<? for this sort, and
 - strip the common prefix from filenames.

 The result is often more concise than the original and helps to avoid
 broken hyperlinks in the summary report.

burgerrg commented on src/swish/profile.ss:
```diff
@@ -464,36 +604,113 @@
+    (fprintf op "span:hover { outline: 1px solid darkgray; }\n")
+    (fprintf op "table.code { padding:0; border-spacing:0; white-space: pre; }\n")
+    (fprintf op "table.code td { padding-top:0; padding-bottom:0; }\n")
+    ;; For convenince when copy/pasting from the rendered output, the
+    ;; line numbers should not be selectable.
```
Nice idea!

burgerrg replied:
`convenince` => `convenience`

owaddell replied:
I agree. Well done, Chris!

burgerrg reviewed (approved):
I like the new style!

burgerrg commented:
According to the HTML5 validator:

- The type attribute for the style element is not needed and should be omitted.
- The summary attribute on the table element is obsolete. Consider describing the structure of the table in a caption element or in a figure element containing the table; or, simplify the structure of the table so that no description is needed.

payneca committed:
html: add lang="en" if not otherwise specified

Update our (html5 ...) form to add a lang attribute if none is specified.
See https://www.w3.org/International/techniques/authoring-html.en#language

This also keeps Firefox from suggesting translations for our profile
summary reports.

owaddell commented:
We made some updates to address HTML5 validator warnings. Turns out they have a container image that makes it a bit handier. There is a CSS warning that looks like a known bug in the validator.

We also ran HTML Tidy 5.8.0.

burgerrg reviewed (approved):
Nice work!

owaddell commented:
Thanks for the reviews, Bob!
